xen/arch/arm/mm.c

```diff
@@ -1287,24 +1287,61 @@ long arch_memory_op(int op, XEN_GUEST_HANDLE_PARAM(void) arg)
         if ( d == NULL )
             return -ESRCH;
 
-        for ( i = 0; i < req.num_frames; i++ )
+        if ( likely (req.num_frames <= 32) )
         {
-            xen_pfn_t pa, ma;
+            /* more than 95% cases for GSX operation */
+            xen_pfn_t pma[req.num_frames];
 
-            if ( unlikely(copy_from_guest_offset(&pa, req.pa, i, 1)) )
+            if ( unlikely(copy_from_guest(pma, req.pa, req.num_frames)) )
             {
                 rcu_unlock_domain(d);
                 return -EFAULT;
             }
 
-            ma = pfn_to_paddr(mfn_x(p2m_lookup(d, _gfn(paddr_to_pfn(pa)), NULL)));
+            for ( i = 0; i < req.num_frames; i++ )
+            {
+                pma[i] = pfn_to_paddr(mfn_x(gfn_to_mfn(d, _gfn(paddr_to_pfn(pma[i])))));
+            }
```

Comment:
I have a feeling that such code might be moved to a separate helper. No?

Reply:
done